proposals/unions.md

```diff
@@ -0,0 +1,322 @@
+* Otherwise, the pattern is applied to the result of accessing the `IUnion.Value` property on the incoming union.
+
+#### Union exhaustiveness
+
+A union type is assumed to be "exhausted" by its case types. This means that a `switch` expression is exhaustive if it handles all of a union's case types:
+
+
+``` c#
+var name = pet switch
+{
+    Dog dog => ...,
+    Cat cat => ...,
+    // No warning about non-exhaustive switch
+};
+```
+
+#### Nullability
+
+The null state of a union's `Value` property is tracked like any other property, with these modifications: 
+
+- For union types where none of the case types are nullable, the default state for `Value` is "not null" rather than "maybe null". 
+- When a union constructor is called (explicitly, through a union conversion or via an implementation of `IUnion<TUnion>.TryCreate(...)`), the new union's `Value` gets the null state of the incoming value.
+- When the non-boxing access pattern's `HasValue` or `TryGetValue(...)` are used to query the contents of a union type (explicitly or via pattern matching), it impacts `Value`'s nullability state in the same way as if `Value` had been checked directly: The null state of `Value` becomes "not null" on the `true` branch.
+
+Even when a union switch is otherwise exhaustive, if the null state of the incoming union's `Value` property is "maybe null", a warning will be given on unhandled null.
+
+``` c#
+Pet pet = GetNullableDog(); // 'pet.Value' is "maybe null"
+var value = pet switch
+{
+    Dog dog => ...,
+    Cat cat => ...,
+    // Warning: 'null' not handled
+}
+```
+
+### Union declarations
+
+Union declarations are a succinct and opinionated way of declaring union types in C#. They declare a struct which uses a single object reference for storage, which means that:
+
+* Any value types among their case types will be boxed on entry.
+* Union values are pointer-length in size, and thus protected from "tearing".
+
+The intent is for union declarations to cover the vast majority of use cases quite nicely. The two main reasons for hand coding specific union types rather than use union declarations are expected to be:
+
+* Adapting existing types to the union patterns to gain union behaviors.
+* Implementing a different storage strategy for e.g. efficiency or interop reasons.
+
+#### Syntax
+
+A union declaration has a name and a list of case types.
+
+``` antlr
+union_declaration
+    : attributes? struct_modifier* 'partial'? 'union' identifier type_parameter_list?
+      '(' type (',' type)* ')' type_parameter_constraints_clause* 
+      (`{` struct_member_declaration* `}` | ';')
+    ;
+```
+
+In addition to the restrictions on struct members ([§16.3](https://github.com/dotnet/csharpstandard/blob/draft-v8/standard/structs.md#163-struct-members)), the following applies to union members:
+
+* Instance fields, auto-properties or field-like events are not permitted.
+* Explicitly declared public constructors with a single parameter are not permitted.
```

Comment:
What about with no parameters, since the lowered `TryCreate` member is calling the parameterless constructor in order to represent a "null" case?

Reply:
Still though, on whatever principle we block defining the single-param constructors, shouldn't we block defining the "null case" parameter?

Reply:
I took a different action from this, which is to add to the well-formedness rules a requirement that a nullary constructor causes `Value` to be `null`.